File: implementations/csharp/client/tests/Specus.Client.Tests/TurnLongTermAuthenticatorTests.cs
using System.Net;
using System.Security.Cryptography;
using System.Text;
using Specus.Client.Configuration;
using Specus.Client.PeerMesh;

namespace Specus.Client.Tests;

public sealed class TurnLongTermAuthenticatorTests
{
    private const string Username = "1900000000:client-a:01020304";
    private const string Credential = "turn-credential";
    private const string Realm = "specus";
    private const string Nonce = "login-nonce";

    [Theory]
    [InlineData(StunMessage.AllocateRequest)]
    [InlineData(StunMessage.RefreshRequest)]
    [InlineData(StunMessage.CreatePermissionRequest)]
    public void SignsEveryAuthenticatedTurnRequestType(ushort type)
    {
        var authenticator = Authenticator();
        var transactionId = TransactionId(type);
        var request = type switch
        {
            StunMessage.AllocateRequest => StunMessage.Of(
                type,
                transactionId,
                StunMessage.RequestedUdpTransportAttribute()),
            StunMessage.RefreshRequest => StunMessage.Of(
                type,
                transactionId,
                StunMessage.Lifetime(300)),
            _ => StunMessage.Of(
                type,
                transactionId,
                StunMessage.XorPeerAddress(new IPEndPoint(IPAddress.Parse("192.0.2.30"), 3478), transactionId)),
        };

        var encoded = authenticator.Encode(request);
        var parsed = StunMessage.Parse(encoded)!;

        Assert.Equal(Username, parsed.TextAttribute(StunMessage.AttrUsername));
        Assert.Equal(Realm, parsed.TextAttribute(StunMessage.AttrRealm));
        Assert.Equal(Nonce, parsed.TextAttribute(StunMessage.AttrNonce));
        Assert.NotNull(parsed.First(StunMessage.AttrMessageIntegrity));
        Assert.True(VerifyMessageIntegrity(encoded, LongTermKey(Realm)));
    }

    [Fact]
    public void BindingAndSendIndicationRemainUnsigned()
    {
        var authenticator = Authenticator();
        var bindingTransaction = TransactionId(4);
        var indicationTransaction = TransactionId(5);
        var requests = new[]
        {
            StunMessage.Of(
                StunMessage.BindingRequest,
                bindingTransaction,
                StunMessage.Software("test")),
            StunMessage.Of(
                StunMessage.SendIndication,
                indicationTransaction,
                StunMessage.XorPeerAddress(
                    new IPEndPoint(IPAddress.Parse("192.0.2.31"), 3478),
                    indicationTransaction),
                StunMessage.Data([1, 2, 3])),
        };

        foreach (var request in requests)
        {
            var encoded = authenticator.Encode(request);
            var parsed = StunMessage.Parse(encoded)!;

            Assert.Equal(request.ToBytes(), encoded);
            Assert.Null(parsed.First(StunMessage.AttrUsername));
            Assert.Null(parsed.First(StunMessage.AttrRealm));
            Assert.Null(parsed.First(StunMessage.AttrNonce));
            Assert.Null(parsed.First(StunMessage.AttrMessageIntegrity));
        }
    }

    [Fact]
    public void AppliesUnauthorizedAndStaleNonceChallengesButIgnoresOtherErrors()
    {
        var authenticator = Authenticator();
        Assert.False(authenticator.ApplyChallenge(StunMessage.Of(
            StunMessage.AllocateError,
            TransactionId(6),
            StunMessage.ErrorCode(400, "bad-request"),
            StunMessage.Realm("ignored-realm"),
            StunMessage.Nonce("ignored-nonce"))));

        foreach (var code in new[] { 401, 438 })
        {
            var challengedRealm = $"challenge-realm-{code}";
            var challengedNonce = $"challenge-nonce-{code}";
            Assert.True(authenticator.ApplyChallenge(StunMessage.Of(
                StunMessage.AllocateError,
                TransactionId(code),
                StunMessage.ErrorCode(code, "challenge"),
                StunMessage.Realm(challengedRealm),
                StunMessage.Nonce(challengedNonce))));

            var encoded = authenticator.Encode(StunMessage.Of(
                StunMessage.AllocateRequest,
                TransactionId(code + 1),
                StunMessage.RequestedUdpTransportAttribute()));
            var parsed = StunMessage.Parse(encoded)!;
            Assert.Equal(challengedRealm, parsed.TextAttribute(StunMessage.AttrRealm));
            Assert.Equal(challengedNonce, parsed.TextAttribute(StunMessage.AttrNonce));
            Assert.True(VerifyMessageIntegrity(encoded, LongTermKey(challengedRealm)));
        }
    }

    [Fact]
    public void IncompleteCredentialsLeaveProtectedRequestUnsigned()
    {
        var authenticator = new TurnLongTermAuthenticator();
        Assert.True(authenticator.Update(new PeerMeshConfig
        {
            IceUsername = Username,
            IceCredential = Credential,
        }));
        Assert.False(authenticator.CanAuthenticate);

        var request = StunMessage.Of(
            StunMessage.AllocateRequest,
            TransactionId(7),
            StunMessage.RequestedUdpTransportAttribute());
        var parsed = StunMessage.Parse(authenticator.Encode(request))!;

        Assert.Null(parsed.First(StunMessage.AttrMessageIntegrity));
        Assert.Null(parsed.First(StunMessage.AttrUsername));
    }

    private static TurnLongTermAuthenticator Authenticator()
    {
        var authenticator = new TurnLongTermAuthenticator();
        Assert.True(authenticator.Update(Config(Realm, Nonce)));
        Assert.True(authenticator.CanAuthenticate);
        return authenticator;
    }

    internal static PeerMeshConfig Config(string realm, string nonce) => new()
    {
        IceUsername = Username,
        IceCredential = Credential,
        IceRealm = realm,
        IceNonce = nonce,
    };

    internal static byte[] LongTermKey(string realm) =>
        MD5.HashData(Encoding.UTF8.GetBytes($"{Username}:{realm}:{Credential}"));

    internal static bool VerifyMessageIntegrity(byte[] packet, byte[] key)
    {
        var parsed = StunMessage.Parse(packet);
        var expected = parsed?.First(StunMessage.AttrMessageIntegrity)?.Value;
        if (expected is not { Length: 20 } || packet.Length < 24)
        {
            return false;
        }
        using var hmac = new HMACSHA1(key);
        var actual = hmac.ComputeHash(packet, 0, packet.Length - 24);
        return CryptographicOperations.FixedTimeEquals(expected, actual);
    }

    internal static byte[] TransactionId(int seed) =>
        Enumerable.Range(0, StunMessage.TransactionIdBytes).Select(index => (byte)(seed + index)).ToArray();
}

File: implementations/java/client/src/test/java/com/theshuai/specusclient/peer/TurnLongTermAuthenticatorTests.java
package com.theshuai.specusclient.peer;

import com.theshuai.common.clientauth.ClientAuthLoginResponse;
import com.theshuai.common.stun.StunMessage;
import org.junit.jupiter.api.Test;

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class TurnLongTermAuthenticatorTests {
    private static final String USERNAME = "1900000000:client-a:01020304";
    private static final String CREDENTIAL = "turn-credential";
    private static final String REALM = "specus";
    private static final String NONCE = "login-nonce";

    @Test
    void signsAllAuthenticatedTurnRequestTypesWithLongTermKey() throws Exception {
        TurnLongTermAuthenticator authenticator = authenticator();
        InetSocketAddress peer = new InetSocketAddress(InetAddress.getByName("192.0.2.30"), 3478);

        StunMessage[] requests = {
                StunMessage.of(
                        StunMessage.ALLOCATE_REQUEST,
                        transactionId(1),
                        StunMessage.requestedUdpTransportAttribute()),
                StunMessage.of(
                        StunMessage.REFRESH_REQUEST,
                        transactionId(2),
                        StunMessage.lifetime(300)),
                StunMessage.of(
                        StunMessage.CREATE_PERMISSION_REQUEST,
                        transactionId(3),
                        StunMessage.xorPeerAddress(peer, transactionId(3)))
        };

        for (StunMessage request : requests) {
            byte[] encoded = authenticator.encode(request);
            StunMessage parsed = StunMessage.parse(encoded, 0, encoded.length);

            assertEquals(USERNAME, parsed.username().orElseThrow());
            assertEquals(REALM, parsed.realm().orElseThrow());
            assertEquals(NONCE, parsed.nonce().orElseThrow());
            assertTrue(parsed.first(StunMessage.ATTR_MESSAGE_INTEGRITY).isPresent());
            assertTrue(StunMessage.verifyMessageIntegrity(encoded, 0, encoded.length, longTermKey(REALM)));
        }
    }

    @Test
    void bindingAndSendIndicationRemainUnsigned() throws Exception {
        TurnLongTermAuthenticator authenticator = authenticator();
        InetSocketAddress peer = new InetSocketAddress(InetAddress.getByName("192.0.2.31"), 3478);
        byte[] bindingTx = transactionId(4);
        byte[] indicationTx = transactionId(5);
        StunMessage binding = StunMessage.of(
                StunMessage.BINDING_REQUEST,
                bindingTx,
                StunMessage.software("test"));
        StunMessage indication = StunMessage.of(
                StunMessage.SEND_INDICATION,
                indicationTx,
                StunMessage.xorPeerAddress(peer, indicationTx),
                StunMessage.data(new byte[]{1, 2, 3}));

        assertArrayEquals(binding.toBytes(), authenticator.encode(binding));
        assertArrayEquals(indication.toBytes(), authenticator.encode(indication));
        assertUnsigned(authenticator.encode(binding));
        assertUnsigned(authenticator.encode(indication));
    }

    @Test
    void appliesUnauthorizedAndStaleNonceChallengesButIgnoresOtherErrors() throws Exception {
        TurnLongTermAuthenticator authenticator = authenticator();

        assertFalse(authenticator.applyChallenge(StunMessage.of(
                StunMessage.ALLOCATE_ERROR,
                transactionId(6),
                StunMessage.errorCode(400, "bad-request"),
                StunMessage.realm("ignored-realm"),
                StunMessage.nonce("ignored-nonce"))));

        for (int code : new int[]{401, 438}) {
            String challengedRealm = "challenge-realm-" + code;
            String challengedNonce = "challenge-nonce-" + code;
            assertTrue(authenticator.applyChallenge(StunMessage.of(
                    StunMessage.ALLOCATE_ERROR,
                    transactionId(code),
                    StunMessage.errorCode(code, "challenge"),
                    StunMessage.realm(challengedRealm),
                    StunMessage.nonce(challengedNonce))));

            StunMessage request = StunMessage.of(
                    StunMessage.ALLOCATE_REQUEST,
                    transactionId(code + 1),
                    StunMessage.requestedUdpTransportAttribute());
            byte[] encoded = authenticator.encode(request);
            StunMessage parsed = StunMessage.parse(encoded, 0, encoded.length);
            assertEquals(challengedRealm, parsed.realm().orElseThrow());
            assertEquals(challengedNonce, parsed.nonce().orElseThrow());
            assertTrue(StunMessage.verifyMessageIntegrity(
                    encoded, 0, encoded.length, longTermKey(challengedRealm)));
        }
    }

    private TurnLongTermAuthenticator authenticator() {
        ClientAuthLoginResponse.PeerMeshConfig config = new ClientAuthLoginResponse.PeerMeshConfig();
        config.setIceUsername(USERNAME);
        config.setIceCredential(CREDENTIAL);
        config.setIceRealm(REALM);
        config.setIceNonce(NONCE);
        TurnLongTermAuthenticator authenticator = new TurnLongTermAuthenticator();
        assertTrue(authenticator.update(config));
        assertTrue(authenticator.canAuthenticate());
        return authenticator;
    }

    private void assertUnsigned(byte[] packet) {
        StunMessage parsed = StunMessage.parse(packet, 0, packet.length);
        assertFalse(parsed.first(StunMessage.ATTR_MESSAGE_INTEGRITY).isPresent());
        assertFalse(parsed.username().isPresent());
        assertFalse(parsed.realm().isPresent());
        assertFalse(parsed.nonce().isPresent());
    }

    private byte[] longTermKey(String realm) throws Exception {
        return MessageDigest.getInstance("MD5").digest(
                (USERNAME + ":" + realm + ":" + CREDENTIAL).getBytes(StandardCharsets.UTF_8));
    }

    private byte[] transactionId(int seed) {
        byte[] value = new byte[StunMessage.TRANSACTION_ID_BYTES];
        for (int i = 0; i < value.length; i++) {
            value[i] = (byte) (seed + i);
        }
        return value;
    }
}
